manual/embedded-examples/src/docs/dev/procedures.asciidoc

```diff
@@ -0,0 +1,495 @@
+     * with two notable differences. One, it is annotated with {@link PerformsWrites},
+     * which is <i>required</i> if you want to perform updates to the graph in your
+     * procedure.
+     *
+     * Two, it returns {@code void} rather than a stream. This is simply a short-hand
+     * for saying our procedure always returns an empty stream of empty records.
+     *
+     * @param nodeId the id of the node to index
+     * @param propKeys a list of property keys to index, only the ones the node
+     *                 actually contains will be added
+     */
+    @Procedure
+    @PerformsWrites
+    public void index( @Name("nodeId") long nodeId,
+                       @Name("properties") List<String> propKeys )
+    {
+        Node node = db.getNodeById( nodeId );
+
+        // Load all properties for the node once and in bulk,
+        // the resulting set will only contain those properties in `propKeys`
+        // that the node actually contains.
+        Set<Map.Entry<String,Object>> properties =
+                node.getProperties( propKeys.toArray( new String[0] ) ).entrySet();
+
+        // Index every label (this is just as an example, we could filter which labels to index)
+        for ( Label label : node.getLabels() )
+        {
+            Index<Node> index = db.index().forNodes( indexName( label.name() ), FULL_TEXT );
+
+            // In case the node is indexed before, remove all occurrences of it so
+            // we don't get old or duplicated data
+            index.remove( node );
+
+            // And then index all the properties
+            for ( Map.Entry<String,Object> property : properties )
+            {
+                index.add( node, property.getKey(), property.getValue() );
+            }
+        }
+    }
+
+
+    /**
+     * This is the output record for our search procedure. All procedures
+     * that return results return them as a Stream of Records, where the
+     * records are defined like this one - customized to fit what the procedure
+     * is returning.
+     *
+     * These classes can only have public non-final fields, and the fields must
+     * be one of the following types:
+     *
+     * <ul>
+     *     <li>{@link String}</li>
+     *     <li>{@link Long} or {@code long}</li>
+     *     <li>{@link Double} or {@code double}</li>
+     *     <li>{@link Number}</li>
+     *     <li>{@link Boolean} or {@code boolean}</li>
+     *     <li>{@link org.neo4j.graphdb.Node}</li>
+     *     <li>{@link org.neo4j.graphdb.Relationship}</li>
+     *     <li>{@link org.neo4j.graphdb.Path}</li>
+     *     <li>{@link java.util.Map} with key {@link String} and value {@link Object}</li>
+     *     <li>{@link java.util.List} of elements of any valid field type, including {@link java.util.List}</li>
+     *     <li>{@link Object}, meaning any of the valid field types</li>
+     * </ul>
+     */
+    public static class SearchHit
+    {
+        // This records contain a single field named 'nodeId'
```

Comment:
Indeed, but that section comes from somewhere else so I'll see it is fixed there (and it will be 'included' here later).